### ISSUE TITLE
Using DHCP on CoreOS

### ISSUE THREAD

**Issue by helander** [jpetazzo/pipework#20]
pipework uses udhcpc for DHCP leases, but udhcpc does not seem to be available in CoreOs. Anyone that has experience with pipework on CoreOS and use of DHCP?

**Comment by cameron** [comment]
@seblu I await your arrival with baited breath.

**Comment by starkovv** [comment]
+1

**Comment by starkovv** [comment]
@jpetazzo @seblu Any news on topic? :-)

**Comment by flipchart** [comment]
+1

**Comment by jpetazzo** [comment]
My last changes hopefully address this issue: 19bc094

Can you test it and report?

**Comment by jpetazzo** [comment]
Closing older issues.

**Comment by pppq** [comment]
On CoreOS beta 877.1.0 (docker version 1.9.1), the `dhcp` option doesn't seem to work for me. The helper container stays around indefinitely, with the following output:
```
1/2/2016 8:39:50 PM udhcpc (v1.24.1) started
1/2/2016 8:39:50 PM udhcpc: sendto: Network is down
1/2/2016 8:39:50 PM Sending discover...
1/2/2016 8:39:50 PM Read error: Network is down, reopening socket
1/2/2016 8:39:53 PM Sending discover...
1/2/2016 8:39:53 PM udhcpc: sendto: Network is down
1/2/2016 8:39:53 PM Read error: Network is down, reopening socket
1/2/2016 8:39:56 PM Sending discover...
1/2/2016 8:39:56 PM Read error: Network is down, reopening socket
1/2/2016 8:39:56 PM udhcpc: sendto: Network is down
1/2/2016 8:39:59 PM Read error: Network is down, reopening socket
1/2/2016 8:40:02 PM Read error: Network is down, reopening socket
1/2/2016 8:40:05 PM Read error: Network is down, reopening socket
1/2/2016 8:40:08 PM Read error: Network is down, reopening socket
1/2/2016 8:40:11 PM Read error: Network is down, reopening socket
1/2/2016 8:40:14 PM Read error: Network is down, reopening socket
```
The `eth1` interface added by pipework is not in `UP` state by default, and can not be brought up in the container in question:
```
core@coreos ~ $ sudo ./pipework enp0s3 -i eth1 5d489f03addf dhcp
```
```
root@5d489f03addf:/# ifconfig -a
eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:00:03
          inet addr:172.17.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:49 errors:0 dropped:0 overruns:0 frame:0
          TX packets:47 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:3645 (3.6 KB)  TX bytes:2802 (2.8 KB)

eth1      Link encap:Ethernet  HWaddr f2:d4:fe:cd:8b:fb
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:75 errors:0 dropped:0 overruns:0 frame:0
          TX packets:75 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:38281 (38.2 KB)  TX bytes:38281 (38.2 KB)

root@5d489f03addf:/# ifconfig eth1 up
SIOCSIFFLAGS: Operation not permitted
```
It is possible to do so in the DHCP helper container, however:
```
/ # ifconfig eth1 up
```
Following which, the next discovery attempt succeeds, but the interface is never updated with an IP address:
```
1/2/2016 9:26:17 PM Read error: Network is down, reopening socket
1/2/2016 9:26:20 PM Read error: Network is down, reopening socket
1/2/2016 9:26:37 PM Sending discover...
1/2/2016 9:26:40 PM Sending discover...
1/2/2016 9:26:40 PM Sending select for 192.168.178.225...
1/2/2016 9:26:40 PM Lease of 192.168.178.225 obtained, lease time 43200
```
```
root@5d489f03addf:/# ifconfig -a
[...]
eth1      Link encap:Ethernet  HWaddr f2:d4:fe:cd:8b:fb
          inet6 addr: fe80::f0d4:feff:fecd:8bfb/64 Scope:Link
          # no "inet addr" line appears here
          *UP* BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:10 errors:0 dropped:0 overruns:0 frame:0
          TX packets:11 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:1474 (1.4 KB)  TX bytes:1680 (1.6 KB)
[...]
```
The output for `eth1` is the same in the helper container, if I keep `udhcpc` around with the foreground `-f` option.

(Note that `dhcpcd` is installed by default on this version of CoreOS, and pipework succeeds in allocating an address without problems if I use that option instead of `dhcp`.)

**Comment by pppq** [comment]
BTW, this looks very similar to what @dreamcat4 describes in https://github.com/jpetazzo/pipework/issues/163#issuecomment-144538955 and below.

**Comment by pppq** [comment]
For future reference, I'll also mention https://github.com/jpetazzo/pipework/issues/182 here (`dhcpcd` runs indefinitely, because it does not treat option `-q` as "run in one-shot mode").

**Comment by jpetazzo** [comment]
I've merged #186, does it provide alternate DHCP options that could work for you?

**Comment by pppq** [comment]
I am not using pipework actively at the moment, just happened to try it on containers running on CoreOS at the time. To summarize (hopefully correctly):

* While PR #186 does not touch `dhcpcd` options, according to the manual page mentioned in #182, it treats the default `-q` option differently as the other implementations ("keep quiet" instead of "exit immediately after an address is obtained"), effectively already behaving like the `-f` versions. (You will have to find and signal processes to exit that belonged to containers no longer in use on your own.)

* `dhcp`, the busybox sidekick container option, can also be made to work, if the scripts mentioned in #163 are present.

* `systemd-networkd`, which handles the network setup for the machine running CoreOS came up as a possible alternative in this thread; I didn't check if it would be possible to make it also act on any running container's behalf.

* Finally, @jumanjiman pointed to CoreOS' toolbox in https://github.com/jpetazzo/pipework/issues/20#issuecomment-42187225, which makes it possible to install userland tools from Fedora with full system privileges, so theoretically it should be possible to use any of the other clients as well. I didn't try this one either.
